Type completion of facet types is separate from Identifying

danakj opened the pull request:
Identifying a facet type is an operation on a pair of (self type, facet type). It substitutes that self in as the `Self` of any require declarations in order to form the set of (self type, SpecificInterface) pairs that constitute the requirements of the IdentifiedFacetType. Currently we don't pass around any self type, and assume all require declarations are written against `Self` but this will change in the future.

By contrast, type completion is done in the abstract and does not form specifics for the require declarations. The purpose of type completion is to enumerate the scopes where name lookup can occur and ensure they are completed.

With this change, type completion is:
- No longer built on top of identification for facet types.
- Recursively ensures all `extend` scopes are complete since name lookup can find symbols in them.

We add some test cases that demonstrate consistency between a resolving the specific of a generic class, and a generic interface/constraint, both used in a type position. In all cases, an invalid specific is not materialized for the type completion when the specific's arguments are used in a non-extend context. But they specific is materialized and checked for type completion when in an extend context (extend impl or extend require).

Type completion itself does not need to recurse into named constraints or interfaces as the `extend require` declarations require the type to be complete immediately, just as for `extend impl` in a class.

We had a test (`fail_incomplete_where.carbon`) with `impl as J where .Self impls K` and `J` is incomplete, which used to be diagnosed but no longer is, because we don't require non-extend interfaces to be complete in type completion, nor in identification. The test was trying to test the presence of rewrite constraints though, which it didn't even use. So we remove the diagnostic that we can't hit anymore and replaced it with a TODO, and add a test that should reach that TODO once qualified rewrite constraints work.

danakj committed:
complete-not-identify

danakj commented:
@josh11b This comes from discussion with Richard on Friday, and we think may extend (ha ha) a little on what was proposed in https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p5168.md#proposed-rules since those rules do not mention extend.

Our discussion notes [are here](https://docs.google.com/document/d/1Yt-i5AmF76LSvD4TrWRIAE_92kii6j5yFiW-S7ahzlg/edit?tab=t.1ji9ixn9bbnn#heading=h.kijomnov90rz) and I will be following up to introduce a `self` input for the identification of a facet type, so that we can resolve specifics with a real self for identification. For type completion it is left abstract. This difference doesn't exactly show up too much yet but will be critical for `require T as U` instead of `require Self...` as we will need to replace the `T` with some external self via the specific.

danakj committed:
touchup

danakj committed:
MakeCopyOfSpecificAndAppendSelf

danakj committed:
bit-simpler

danakj committed:
whitespace

danakj committed:
complete-at-require-decl

danakj commented:
This needs a rebase to get tests happy, which I will get to later.

danakj committed:
no-recurse-needed

josh11b reviewed (approved):
This PR is fine, but it seems like we should update the official rules with a new proposal.

danakj committed:
recurse-on-facets

danakj commented:
> This PR is fine, but it seems like we should update the official rules with a new proposal.

I had tried to get rid of the recursion onto the facet types in require decls, as we check them for completeness where they are written. However when the interface is generic, the facet type can contain generic bindings in it, and those may have been OK as symbolics, but produce an error with a given specific. We have tests for these included in the PR, which removing the recursion broke.

So I have re-introduced the recursion (via worklist) on the specific facet type in an `extend require` decl, in order to:
- Complete a specific for that facet type, in order to find errors.
- Then look for more `extern require` decls in the interfaces/constraints that it names.

We only need to do this step of forming specifics and recursing into the `extend require` decls if the containing interface/constraint has a generic id, otherwise there are no generic bindings to apply a specific to.

danakj committed:
autoupdate

danakj committed:
comment

danakj committed:
rewrites-complete

danakj commented:
We had a test with `impl as J where .Self impls K` and `J` is incomplete, which used to be diagnosed but no longer is, because we don't require non-extend interfaces to be complete in type completion, nor in identification. The test was trying to test the presence of _rewrite_ constraints though, which it didn't even use. So I have removed the diagnostic that we can't hit anymore and replaced it with a TODO, and added a test that should reach that TODO once qualified rewrite constraints work.

danakj committed:
autoupdate

danakj committed:
avoid-uaf

danakj commented:
The last commit now avoids a UAF which is properly addressed by https://github.com/carbon-language/carbon-lang/pull/6399. I will revert that commit when the fix lands.

danakj committed:
test-and-fix-multi-level-specific-resolution

danakj committed:
one-less-malloc

danakj committed:
add-comment

josh11b commented on toolchain/check/testdata/generic/extend_type_completion.carbon:
```diff
@@ -62,18 +62,24 @@ interface J(N:! i32) {
+}
 
-// J extends K so the type of K is completed, but is invalid.
+// I extend J extends K so the type of K is completed, but is invalid.
```
```suggestion
// I extends J extends K so the type of K is completed, but is invalid.
```

danakj replied:
Done

josh11b commented on toolchain/check/testdata/impl/fail_undefined_interface.carbon:
```diff
@@ -41,6 +41,58 @@ class C {}
+library "[[@TEST_NAME]]";
+
+interface J;
+interface K { let X:! type; }
```
Is this intentional? It is never used.

danakj replied:
No, copy-paste mistake.

danakj committed:
extends

danakj committed:
no-k

danakj committed:
fix-test

danakj reviewed (commented):
PTAL

danakj committed:
Revert "avoid-uaf"

This reverts commit 8c5e57337015aa9cbd40666e07188dafbe0e9546.

zygoloid commented on toolchain/check/testdata/impl/impl_as_named_constraint.carbon:
```diff
@@ -40,17 +40,11 @@ impl () as B {}
   extend require impls A;
 }
 
-// Requries B identified.
-impl () as B;
```
Do we have a test for
```carbon
constraint B;
impl () as B;
```
somewhere? (Which should be an error because `B` is not identified.)

danakj replied:
We do here: https://github.com/carbon-language/carbon-lang/blob/6b1ef75ac56b61d6edc8ecfc32c183d73bb2701e/toolchain/check/testdata/facet/fail_incomplete.carbon#L62

But I have now moved tests around to consolidate things better in impl/incomplete.carbon.

zygoloid commented on toolchain/check/impl_lookup.cpp:
```diff
@@ -211,9 +211,14 @@ static auto GetInterfacesFromConstantId(
+      context.identified_facet_types()
+          .Get(identified_id)
+          .required_interfaces());
+  return {{.interfaces = std::move(interfaces),
```
`SmallVector` moves aren't especially cheap due to needing to copy the inline storage; maybe construct this in-place in `.interfaces` instead to avoid one of the two copies?
```suggestion
  // TODO: IdentifiedFacetTypes are held in a RelationalValueStore which does
  // not protect against UAF through reallocation, so we must copy data out of
  // it.
  return {{.interfaces = context.identified_facet_types()
               .Get(identified_id)
               .required_interfaces(),
```

danakj replied:
This change is reverted in this PR now that #6399 landed.

zygoloid commented on toolchain/check/type_completion.cpp:
```diff
@@ -70,16 +72,143 @@ static auto NoteIncompleteNamedConstraint(
+        specifics.push_back(
+            {extends.specific_id, constraint.require_impls_block_id});
+      }
+    }
```
Would be nice if we could find some way to avoid this duplication, maybe add a TODO?

danakj replied:
I'm not sure. This is emblematic of the duplication between interfaces and named constraints across the toolchain. Since they have different types, with different field names for ids, etc, we end up with a lot of very similar code. I have tried elsewhere to use field pointers and templates but it gets really unreadable really quickly. I don't have any great ideas for how I would address that TODO right now.

zygoloid commented on toolchain/check/type_completion.cpp:
```diff
@@ -70,16 +72,143 @@ static auto NoteIncompleteNamedConstraint(
+          }
+          work.push_back(facet_type_id);
+        }
+      }
```
I think we may be able to avoid this:

In `handle_require.cpp`, move the `RequireCompleteType` call for an `extend require impls` declaration later, to after we finish the generic for the `RequireImplsDecl`.

That way, if the enclosing context is itself generic, then we'll create a `require_complete_type` instruction in the generic for the interface / named constraint, which will cause the `ResolveSpecificDefinition` below to recursively require the extended type to be complete. (That's how we get recursive completeness requirements for `extend` in classes.)

danakj replied:
Trivially making this change did not work:
```
CHECK failure at toolchain/sem_ir/generic.cpp:96: specific_ir.generics().GetSelfSpecific( specific_ir.specifics().Get(specific_id).generic_id) == specific_id: Queried generic_inst_in_decl0 in specific5800002B for {kind: NamedConstraintDecl, arg0: constraint58000000, arg1: inst_block5800001B, type: type(inst58000041)} before it was resolved.
```

This looks like when I tried to resolve an inner specific facet type before resolving the containing constraint's specific. I think this strategy maybe just works for constraint->require but does not work for constraint->require->constraint->require as there's nothing crossing that middle require->constraint boundary.

I can try understand this a bit better tomorrow.

danakj replied:
Ok I found one issue.

ResolveSpecificDeclForInst() in eval was not resolving specific decl blocks for named constraints in a facet type after applying a specific.

But fixing that turned up a crash in import instead (that I can reproduce on interfaces too without that fix).

For the record this crash was fixed by Jon in #6409.

danakj replied:
Alright, this is done. I apologize cuz I have rebased this a few times to make it work since HEAD had fixes I needed, and mixing commits with merges makes for real confusion (for me?).

danakj committed:
move-tests

danakj committed:
try-move-requirecompletetype

danakj committed:
works-but-import-crash

danakj committed:
wip

danakj committed:
works

danakj committed:
yay

danakj committed:
remove-todo-diagnostic

danakj commented on toolchain/check/facet_type.cpp:
```diff
@@ -62,12 +62,10 @@ static auto IncompleteFacetTypeDiagnosticBuilder(
-                                   facet_type_inst_id);
+    context.TODO(loc_id,
+                 "declaration of impl as incomplete interface with a rewrite "
+                 "constraint");
```
I am quite convinced this case should no longer ever happen so I changed it into a CHECK with a TODO to remove the parameter. I tried to organize the type completion of the facet type in `impl as` a bit better since it occurs in two different places with some loose coordination right now, but it's very tricky. I will try again elsewhere.

danakj replied:
Ok yeah, path entirely goes away in https://github.com/carbon-language/carbon-lang/pull/6420

zygoloid commented on toolchain/check/testdata/impl/incomplete.carbon:
```diff
@@ -0,0 +1,862 @@
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/impl/incomplete.carbon
+
+// --- fail_empty_struct.carbon
+library "[[@TEST_NAME]]";
+
+interface I;
+
+// Requires I identified.
+impl {} as I;
+
```
```suggestion
// --- forward_decl.carbon
library "[[@TEST_NAME]]";

interface I;

// Requires I identified.
impl {} as I;

// --- fail_empty_struct.carbon
library "[[@TEST_NAME]]";

interface I;

impl {} as I;

```
... so that we have a non `fail_` test for forward-declaring an impl of a forward-declared interface.

danakj replied:
If we split it, the first test will fail with "impl declaration was never defined", so it still ends up being a fail_ test.

zygoloid replied:
Aah fair enough, never mind then.

zygoloid commented on toolchain/check/facet_type.cpp:
```diff
@@ -55,20 +55,15 @@ static auto WitnessQueryMatchesInterface(
-    return context.emitter().Build(loc_id, ImplAsIncompleteFacetTypeRewrites,
-                                   facet_type_inst_id);
-  }
+  // TODO: Remove this parameter. Facet types don't need to be complete for in
```
```suggestion
  // TODO: Remove this parameter. Facet types don't need to be complete for impl
```

danakj replied:
Done

zygoloid commented on toolchain/check/handle_require.cpp:
```diff
@@ -271,8 +275,28 @@ auto HandleParseNode(Context& context, Parse::RequireDeclId node_id) -> bool {
+                                             RequireImplsIncompleteFacetType,
+                                             constraint_inst_id);
+            })) {
+      // DiscardGenericDecl(context);
```
```suggestion
```

danakj replied:
Oops yes thanks.

danakj committed:
review
